File: Core/Entities/Order/Order.cs
namespace Core.Entities.Order;

public class Order : BaseEntity
{
    public Order() { }
    public Order(
        string buyerEmail, OrderedAddress address,
        DeliveryMethod deliveryMethod, IReadOnlyList<OrderItem> items,
        decimal subtotal, string paymentIntentId)
    {
        BuyerEmail = buyerEmail;
        Address = address;
        DeliveryMethod = deliveryMethod;
        Items = items;
        Subtotal = subtotal;
        PaymentIntentId = paymentIntentId;
    }

    public string BuyerEmail { get; set; }
    public OrderedAddress Address { get; set; }
    public DeliveryMethod DeliveryMethod { get; set; }
    public IReadOnlyList<OrderItem> Items { get; set; }
    public OrderStatus Status { get; set; } = OrderStatus.Pending;
    public string PaymentIntentId { get; set; }
    public decimal Subtotal { get; set; }
    public decimal GetTotal()
    {
        return Subtotal + DeliveryMethod.Price;
    }
}

File: Client/src/app/types/entities/order.ts
import { IAddress } from './address'
import { BaseEntity } from './base-entity'
import { IDeliveryMethod } from './delivery-method'
import { IOrderItem } from './order-item'

type OrderedAddress = Omit<IAddress, 'id' | 'createdAt'>

export enum OrderStatus {
  Pending = 'Pending',
  PaymentReceived = 'Payment Received',
  PaymentFailed = 'Payment Failed',
}

export interface IOrder extends BaseEntity<number> {
  buyerEmail: string
  address: OrderedAddress
  deliveryMethod: IDeliveryMethod
  items: IOrderItem[]
  status: OrderStatus
  paymentId: string
  subtotal: number
  total: number
}

export interface ICreateOrder {
  buyerEmail: string
  address: OrderedAddress
  deliveryMethodId: number
  basketId: string
}
